internal/db/collection_retriever.go

```diff
@@ -0,0 +1,72 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package db
+
+import (
+	"context"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/internal/db/description"
+)
+
+// CollectionRetriever is a helper struct that retrieves a collection from a document ID.
+type CollectionRetriever struct {
+	db *db
+}
+
+// NewCollectionRetriever creates a new CollectionRetriever.
+func NewCollectionRetriever(database client.DB) *CollectionRetriever {
+	internalDB, ok := database.(*db)
+	if !ok {
```

Comment:
guessing the linter will complain as all panics need to now be documented by putting under a linter suppressed `must` function, or just a linter ignore comment.

Reply:
Can we remove that rule?  That sounds like a bad rule to have

Reply:
When was it added?

Reply:
Was added here: https://github.com/sourcenetwork/defradb/pull/3134

Somewhat related discussion: https://github.com/sourcenetwork/defradb/pull/3134#discussion_r1802084140

Reply:
> Can we remove that rule? That sounds like a bad rule to have

I am happy with the consensus if people think it's not useful then so be it.

Just to understand your position you are saying implicit panics should be allowed?:

`internalDB := database.(*db)`

Which means we shouldn't enforce the users to either check the `ok` or suppress linter with a comment explaining why we are okay if it panics?:

`internalDB, ok := database.(*db)`

Reply:
Ah okay - cheers - maybe we should discuss this rule some more - here it appears to have encouraged bad behaviour.  Although the code itself (constructor param loss of type safety) is fairly un-ideal anyway.

@islamaliev suggestion: I think we can actually remove the awkwardness of this func anyway, by just using `client.Db` instead of casting - the only func on `db.db` that is used seems to be `getCollectionFromRootSchema`, and that could be replaced by a call to the public `GetCollections` function.

Reply:
> Just to understand your position you are saying implicit panics should be allowed?

I'm not sure on my position, but here the linter arguably damaged the codebase.  That damage was dependent on the somewhat misbehaving function signature though - so you could argue that the linter flagged the smell and it was incorrectly handled by the author.

Reply:
> > Just to understand your position you are saying implicit panics should be allowed?
> 
> I'm not sure on my position, but here the linter arguably damaged the codebase. That damage was dependent on the somewhat misbehaving function signature though - so you could argue that the linter correctly flagged the smell and it was incorrectly handled by the author.

I am in a similar boat regard my position. However I am just guessing that the linter is why islam did this, maybe he had other reasons.